### PR TITLE
feat: Batch multiple payments, with unit-tests

### PR DESCRIPTION
**Pull request by Dadogg80**
## Context:
The purpose of this new feature is to be able to execute a batch of ERC20 transfers/ payments in just one transaction through ERC20FeeProxy.sol. This smart contract is optimized for gas usage, therefore, no usage of the SafeERC20.sol library. 

### Public methods in this contract:
- `batchOrphanERC20Payment` accepts a token address, a list of accounts, and a list of values, corresponding with the accounts list. 
-  `batchERC20PaymentWithReferenceAndFee` accepts a token address, a list of accounts, a list of values, and a list of payment references, corresponding with the other parameters. This method uses `ERC20FeeProxy.sol` to complete the transactions with payment references.

### Unit-Testing:
**Success path**
- `Should execute a batch payment of an ERC20 to four accounts`.
- `Should pay multiple ERC20 payments with paymentRef`.

**Fails path**
- `Should revert without approval`.
- `Should revert if not enough ether value`.
- `Should revert if not enough tokens`.

### CODE REVIEW

File: packages/smart-contracts/src/contracts/BatchPayments.sol [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        for (uint256 i = 0; i < recipients.length; i++)
+            token.transferFrom(msg.sender, recipients[i], values[i]);
+    }
+
+    /// @notice Send a batch of erc20 payments with paymentReferences to multiple accounts.
+    /// @param recipients List of recipients accounts as input.
+    /// @param values List of values, corresponding to payees[] as input.
+    /// @param paymentReferences List of paymentRefs, corresponding to the payees[] and values[].
+    /// @dev This method uses ERC20Proxy.sol to pay an invoice with a paymentReference.
+    function batchERC20PaymentWithReference(
```
Comment (yomarion) [inline packages/smart-contracts/src/contracts/BatchPayments.sol:88]: Fees amounts are missing as well as one Fee address (factorized as well as the token)

File: packages/smart-contracts/src/contracts/BatchPayments.sol [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        
+        emit EthTransfer(msg.sender, msg.value);
+    }
+
+    /// @notice Send a batch of erc20 payments to multiple accounts.
+    /// @param token The token to transact with. 
+    /// @param recipients Takes a list of accounts as input.
+    /// @param values Takes a list of values, corresponding to recipients[] as input.
+    /// @dev Remember to give approval for this contract to spend user tokens.
+    function batchERC20Payment(
```
Comment (yomarion) [inline packages/smart-contracts/src/contracts/BatchPayments.sol:74]:
```suggestion
    function batchOrphanERC20Payments(
```
Suggestion to highlight that there is no payment reference tagging

File: packages/smart-contracts/src/contracts/BatchPayments.sol [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        address _tokenAddress,
+        address _to,
+        uint256 _amount,
+        bytes calldata _paymentReference
+    ) external;
+}
+
+contract BatchPayments {
+    
+    IERC20Proxy public erc20Proxy;
```
Comment (yomarion) [inline packages/smart-contracts/src/contracts/BatchPayments.sol:23]:
```suggestion
    IERC20FeeProxy public erc20Proxy;
```

**Review by yomarion — COMMENTED** [review]
Good stuff

File: packages/smart-contracts/src/contracts/BatchPayments.sol [hunk 4]
```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+interface IERC20Proxy {
```
[the remark this answers is not in the document]
Comment (Dadogg80) [reply packages/smart-contracts/src/contracts/BatchPayments.sol:6]: Done.

File: packages/smart-contracts/test/contracts/BatchPayments.test.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        .batchERC20PaymentWithReference(
+          token.address,
+          [payeeOne, payeeTwo, payeeThree, payeeFour],
+          [20, 20, 20, 20],
+          [referenceExample1, referenceExample2, referenceExample3, referenceExample4],
+        ),
+    )
+      .to.emit(token, 'Transfer')
+      .to.emit(batch, 'TransferWithReference')
+      .withArgs(token.address, payeeTwo, '20', ethers.utils.keccak256(referenceExample2));
```
[the remark this answers is not in the document]
Comment (Dadogg80) [reply packages/smart-contracts/test/contracts/BatchPayments.test.ts]: ok

File: packages/smart-contracts/src/contracts/BatchPayments.sol (hunk 2 above)
Comment (Dadogg80) [reply packages/smart-contracts/src/contracts/BatchPayments.sol]: done

File: packages/smart-contracts/src/contracts/BatchPayments.sol (hunk 3 above)
Comment (Dadogg80) [reply packages/smart-contracts/src/contracts/BatchPayments.sol]: done

File: packages/smart-contracts/src/contracts/BatchPayments.sol [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            "transferFromWithReference(address,address,uint256,bytes)",
+                address(token),
+                recipients[i], 
+                values[i],
+                paymentReferences[i]
+                )
+            );
+        require(status, "transferFromWithReference failed");
+
+        emit TransferWithReference(
```
[the remark this answers is not in the document]
Comment (Dadogg80) [reply packages/smart-contracts/src/contracts/BatchPayments.sol]: removed

File: packages/smart-contracts/src/contracts/BatchPayments.sol [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+           (bool status, ) = address(erc20Proxy).delegatecall(
+            abi.encodeWithSignature(
+            "transferFromWithReference(address,address,uint256,bytes)",
+                address(token),
+                recipients[i], 
+                values[i],
+                paymentReferences[i]
+                )
+            );
+        require(status, "transferFromWithReference failed");
```
[the remark this answers is not in the document]
Comment (Dadogg80) [reply packages/smart-contracts/src/contracts/BatchPayments.sol]: Yes, only the last transaction is reverted, the others are ok, I just had a hard time writing the test for it.

File: packages/smart-contracts/src/contracts/BatchPayments.sol (hunk 1 above)
Comment (Dadogg80) [reply packages/smart-contracts/src/contracts/BatchPayments.sol]: done

**Review by yomarion — COMMENTED** [review]
Looks good, nice work ! I wonder about the use of `TestErc20` for the tests, shouldn't we make the difference between balance and approval errors?

**Pull request closed without merging**